File: riveredge-frontend/src/apps/kuaiplm/services/trial-flow.ts
/**
 * 试流 API（R-08）
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/trial-flows';

export type TrialFlowBusinessType = 'component' | 'structure' | 'complete';
export type TrialFlowStatus =
  | 'draft'
  | 'pending'
  | 'approved'
  | 'in_progress'
  | 'concluded'
  | 'closed';

export interface TrialFlowMaterialLine {
  id?: number;
  uuid?: string;
  line_no?: number;
  material_id?: number | null;
  material_code: string;
  material_name: string;
  qty?: number | string | null;
  unit?: string | null;
  remarks?: string | null;
}

export interface TrialFlowStep {
  id: number;
  uuid: string;
  step_key: string;
  step_name: string;
  dept_code: string;
  sort_order: number;
  status: string;
  result?: string | null;
  result_notes?: string | null;
  step_description?: string | null;
  defect_rate?: number | null;
  filled_by_name?: string | null;
  filled_at?: string | null;
}

export interface TrialFlowFormProfileHeaderField {
  key: string;
  label: string;
  sort?: number;
  type?: string;
  options?: Array<{ value: string; label: string }>;
}

export interface TrialFlowFormProfile {
  industry_profile_enabled?: boolean;
  field_labels: Record<string, string>;
  header_fields: TrialFlowFormProfileHeaderField[];
  step_templates: Record<string, Array<Record<string, unknown>>>;
  validation_rules: Array<Record<string, unknown>>;
}

export interface TrialFlow {
  uuid: string;
  id?: number;
  trial_code: string;
  project_id: number;
  project_code: string;
  project_name: string;
  business_type: TrialFlowBusinessType;
  title: string;
  status: TrialFlowStatus;
  current_step_key?: string | null;
  conclusion?: string | null;
  conclusion_summary?: string | null;
  remarks?: string | null;
  extension_payload?: Record<string, unknown> | null;
  materials?: TrialFlowMaterialLine[];
  steps?: TrialFlowStep[];
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface TrialFlowPayload {
  project_id: number;
  business_type: TrialFlowBusinessType;
  title: string;
  remarks?: string | null;
  extension_payload?: Record<string, unknown> | null;
  materials?: TrialFlowMaterialLine[];
}

function unwrapList(res: unknown): { items: TrialFlow[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as TrialFlow[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const trialFlowApi = {
  formProfile: async () => (await api.get(`${BASE}/meta/form-profile`)) as TrialFlowFormProfile,
  list: async (params: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    business_type?: string;
    project_id?: number;
  }) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => (await api.get(`${BASE}/${id}`)) as TrialFlow,
  create: async (data: TrialFlowPayload) =>
    (await api.post(BASE, data)) as TrialFlow,
  update: async (id: number, data: Partial<TrialFlowPayload>) =>
    (await api.put(`${BASE}/${id}`, data)) as TrialFlow,
  submit: async (id: number) =>
    (await api.post(`${BASE}/${id}/submit`)) as TrialFlow,
  approve: async (id: number) =>
    (await api.post(`${BASE}/${id}/approve`)) as TrialFlow,
  reject: async (id: number) =>
    (await api.post(`${BASE}/${id}/reject`)) as TrialFlow,
  fillStep: async (
    id: number,
    stepKey: string,
    data: {
      result: string;
      result_notes?: string;
      step_description?: string;
      defect_rate?: number;
    },
  ) =>
    (await api.post(`${BASE}/${id}/steps/${stepKey}/fill`, data)) as TrialFlow,
  conclude: async (id: number, data: { conclusion: string; conclusion_summary?: string }) =>
    (await api.post(`${BASE}/${id}/conclude`, data)) as TrialFlow,
  close: async (id: number) =>
    (await api.post(`${BASE}/${id}/close`)) as TrialFlow,
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
};

File: riveredge-backend/src/apps/kuaiplm/schemas/trial_flow.py
"""试流 Schema（R-08）"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class TrialFlowMaterialLineIn(BaseModel):
    material_id: Optional[int] = None
    material_code: str = Field(..., min_length=1, max_length=80)
    material_name: str = Field(..., min_length=1, max_length=200)
    qty: Optional[Decimal] = None
    unit: Optional[str] = Field(None, max_length=20)
    remarks: Optional[str] = Field(None, max_length=500)


class TrialFlowMaterialLineOut(TrialFlowMaterialLineIn):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    line_no: int


class TrialFlowStepOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    step_key: str
    step_name: str
    dept_code: str
    sort_order: int
    status: str
    result: Optional[str] = None
    result_notes: Optional[str] = None
    step_description: Optional[str] = None
    defect_rate: Optional[Decimal] = None
    filled_by: Optional[int] = None
    filled_by_name: Optional[str] = None
    filled_at: Optional[datetime] = None


class TrialFlowCreate(BaseModel):
    project_id: int
    business_type: str = Field(..., description="component/structure/complete")
    title: str = Field(..., min_length=1, max_length=200)
    trial_code: Optional[str] = Field(None, max_length=50)
    remarks: Optional[str] = None
    extension_payload: Optional[dict] = None
    materials: List[TrialFlowMaterialLineIn] = Field(default_factory=list)


class TrialFlowUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    remarks: Optional[str] = None
    extension_payload: Optional[dict] = None
    materials: Optional[List[TrialFlowMaterialLineIn]] = None


class TrialFlowStepFill(BaseModel):
    result: str = Field(..., description="pass/fail/na")
    result_notes: Optional[str] = None
    step_description: Optional[str] = Field(None, description="工序描述")
    defect_rate: Optional[Decimal] = Field(None, description="不良率 0-100")


class TrialFlowConclude(BaseModel):
    conclusion: str = Field(..., description="pass/fail/conditional")
    conclusion_summary: Optional[str] = None


class TrialFlowResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: UUID
    id: int
    trial_code: str
    project_id: int
    project_code: str
    project_name: str
    business_type: str
    title: str
    status: str
    current_step_key: Optional[str] = None
    conclusion: Optional[str] = None
    conclusion_summary: Optional[str] = None
    remarks: Optional[str] = None
    extension_payload: Optional[dict] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    concluded_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    materials: List[TrialFlowMaterialLineOut] = Field(default_factory=list)
    steps: List[TrialFlowStepOut] = Field(default_factory=list)


class TrialFlowListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: UUID
    id: int
    trial_code: str
    project_id: int
    project_code: str
    project_name: str
    business_type: str
    title: str
    status: str
    current_step_key: Optional[str] = None
    conclusion: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class TrialFlowListResponse(BaseModel):
    items: List[TrialFlowListItem]
    total: int


class TrialFlowFormProfile(BaseModel):
    """试流表单/工序链扩展 profile。"""

    industry_profile_enabled: bool = Field(
        False, description="是否已启用行业包 profile（false 时前端保持通用 UI）"
    )
    field_labels: Dict[str, str] = Field(default_factory=dict)
    header_fields: List[Dict[str, Any]] = Field(default_factory=list)
    step_templates: Dict[str, List[Dict[str, Any]]] = Field(default_factory=dict)
    validation_rules: List[Dict[str, Any]] = Field(default_factory=list)
